File: dotnet/Core/OpenStack/Manager.cs
using OpenStack.Gfx;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Numerics;
using System.Threading.Tasks;
using static OpenStack.CellManager;

namespace OpenStack;

#region Database

public interface IDatabase {
    object Convert(object src);
    object Query(object src);
}

public interface ICellDatabase {
    ISource Archive { get; }
    IQuery Query { get; }
    Int3 CellId { get; }
    bool CellInterior { get; }
    Vector3 PlayerPosition { get; }
    Quaternion PlayerRotation { get; }
}

#endregion

#region CellManager

public class CellManager(IQuery query, AsyncCoroutineQueue queue, CellBuilder builder) {
    public class Cell(object cellObj, object objectsObj, ICell record, IAsyncEnumerator<object> task) {
        public object Obj = cellObj;
        public object ObjectsObj = objectsObj;
        public ICell Record = record;
        public IAsyncEnumerator<object> Task = task;
    }

    public class CellRef {
        public ICellXref Obj;
        public object Record;
        public string ModelPath;
    }

    public interface ICellXref {
        string Name { get; }
        float? Scale { get; }
        Vector3 Position { get; }
        Vector3 EulerAngles { get; }
    }

    public interface ICellXrefModel {
        string ModelPath { get; }
    }

    public interface ICell {
        uint Id { get; }
        bool IsInterior { get; }
        Int3 GridId { get; }
        string Name { get; }
        Color? AmbientLight { get; }
        List<ICellXref> Xrefs { get; }
    }

    public interface ILand {
        Int3 GridId { get; }
        uint[] Vtex { get; }
        float HeightOffset { get; }
        sbyte[] Heights { get; }
    }

    public interface ILtex {
        long Intv { get; }
        string Path { get; }
    }

    public interface ILigh {
        float Radius { get; }
        Color LightColor { get; }
    }

    public interface IQueryFunc {
        IQuery GetQuery();
    }

    public interface IQuery {
        float MeterInUnits { get; }
        float CellLengthInMeters { get; }
        int[] Radius { get; }
        int World { get; }
        void SetWorld(int world);
        Int3 GetCellId(Vector3 point);
        object FindAnyByName(string name);
        ICell FindCell(Int3 cell);
        ICell FindCellByName(string name);
        ILand FindLand(Int3 cell);
        ILtex FindLtex(int index);
    }

    public readonly IQuery Query = query;
    protected readonly AsyncCoroutineQueue Queue = queue;
    protected readonly CellBuilder Builder = builder;
    readonly Dictionary<Int3, Cell> Cells = [];
    readonly int Radius = query.Radius[0];
    readonly int Radius2 = query.Radius[1];

    public Cell BeginCell(Int3 point) {
        var record = Query.FindCell(point);
        if (record == null) return null;
        var cell = BuildCell(record); Cells[point] = cell;
        return cell;
    }

    public Cell BeginCellByName(string name) {
        var record = Query.FindCellByName(name);
        if (record == null) return null;
        var cell = BuildCell(record); Cells[Int3.Zero] = cell;
        return cell;
    }

    public async Task UpdateCells(Vector3 position, bool immediate = false, int radius = -1) {
        if (radius < 0) radius = Radius;
        var point = Query.GetCellId(position);
        int minX = point.X - radius, maxX = point.X + radius, minY = point.Y - radius, maxY = point.Y + radius;

        // destroy out of range cells
        var outOfRange = new List<Int3>();
        foreach (var s in Cells.Keys)
            if (s.X < minX || s.X > maxX || s.Y < minY || s.Y > maxY) outOfRange.Add(s);
        foreach (var s in outOfRange) DestroyCell(s);

        // create new cells
        var world = Query.World;
        for (var r = 0; r <= radius; r++)
            for (var x = minX; x <= maxX; x++)
                for (var y = minY; y <= maxY; y++) {
                    var p = new Int3(x, y, world); var d = Math.Max(Math.Abs(point.X - p.X), Math.Abs(point.Y - p.Y));
                    if (d == r && !Cells.ContainsKey(p)) {
                        var cell = BeginCell(p);
                        if (cell != null && immediate) await Queue.WaitFor(cell.Task);
                    }
                }

        // update LODs
        foreach (var (p, cell) in Cells) { var d = Math.Max(Math.Abs(point.X - p.X), Math.Abs(point.Y - p.Y)); Builder.SetVisible(cell.ObjectsObj, d <= Radius2); }
    }

    Cell BuildCell(ICell cell) {
        Debug.Assert(cell != null);
        string cellName;
        ILand land;
        if (!cell.IsInterior) { cellName = $"cell {cell.GridId}"; land = Query.FindLand(cell.GridId); }
        else { cellName = $"cell {cell.Name}"; land = null; }
        var (obj, objectsObj) = Builder.CreateContainers(cellName);
        var task = Builder.Coroutine(cell, land, obj, objectsObj); Queue.Add(task);
        return new Cell(obj, objectsObj, cell, task);
    }

    void DestroyCell(Int3 point) {
        if (Cells.TryGetValue(point, out var s)) { Queue.Cancel(s.Task); Builder.Destroy(s.Obj); Cells.Remove(point); }
        else Log.Error("Tried to destroy a cell that is not created.");
    }

    public void DestroyAllCells() {
        foreach (var s in Cells.Values) { Queue.Cancel(s.Task); Builder.Destroy(s.Obj); }
        Cells.Clear();
    }
}

public abstract class CellBuilder {
    public abstract (object, object) CreateContainers(string name);
    public abstract void SetVisible(object src, bool visible);
    public abstract IAsyncEnumerator<object> Coroutine(ICell cell, ILand land, object obj, object objectsObj);
    public abstract void Destroy(object src);
}

public class CellBuilder<Object, Material, Texture, Shader>(ISource source, IQuery query, IOpenGfx[] gfx) : CellBuilder {
    const string DefaultLandTexturePath = "textures/_land_default.dds";
    protected ISource Source = source;
    protected IQuery Query = query;
    protected IOpenGfxApi<Object, Material> GfxApi = (IOpenGfxApi<Object, Material>)gfx[GfX.XApi];
    protected IOpenGfxModel<Object, Material, Texture, Shader> GfxModel = (IOpenGfxModel<Object, Material, Texture, Shader>)gfx[GfX.XModel];
    protected IOpenGfxLight<Object> GfxLight = (IOpenGfxLight<Object>)gfx[GfX.XLight];
    protected IOpenGfxTerrain<Object, Material, Texture> GfxTerrain = (IOpenGfxTerrain<Object, Material, Texture>)gfx[GfX.XTerrain];
    protected float MeterInUnits = query.MeterInUnits;
    protected float CellLengthInMeters = query.CellLengthInMeters;
    static readonly Dictionary<Texture, GfxTerrainLayer<Texture>> TerrainLayers = [];

    public override (object, object) CreateContainers(string name) {
        var obj = GfxApi.CreateObject(name, "Cell");
        var objectsObj = GfxApi.CreateObject("objects", parent: obj);
        return (obj, objectsObj);
    }

    public override void SetVisible(object src, bool visible) => GfxApi.SetVisible((Object)src, visible);
    public override void Destroy(object src) => GfxApi.Destroy((Object)src);

    /// <summary>
    /// A coroutine that instantiates the terrain for, and all objects in, a cell.
    /// </summary>
    public async override IAsyncEnumerator<object> Coroutine(ICell cell, ILand land, object obj, object objectsObj) {
        if (cell == null && land == null) yield break;
        Object obj2 = (Object)obj, objectsObj2 = (Object)objectsObj;
        var cellRefs = GetCellRefs(cell);
        if (land != null && GfxTerrain != null) { yield return null; await CreateLand(land, obj2); yield return null; }
        foreach (var s in cellRefs) await CreateCell(cell, objectsObj2, s);
        if (GfxLight != null) CreateReflectionProbe(cell, obj2);
    }

    CellRef[] GetCellRefs(ICell cell) => [.. cell.Xrefs.Select(s => {
        var record = Query.FindAnyByName(s.Name);
        var modelPath = record != null && record is ICellXrefModel modl ? modl.ModelPath : null;
        return new CellRef { Obj = s, Record = record, ModelPath = modelPath };
    })];

    /// <summary>
    /// Instantiates an object in a cell. Called by InstantiateCellObjectsCoroutine after the object's assets have been pre-loaded.
    /// </summary>
    async Task CreateCell(ICell cell, Object parent, CellRef r) {
        if (r.Record == null) return; //{ Log.Info($"Unknown Object: {r.Obj.Name}"); return; }
        Object modelObj = default; var obj = r.Obj;
        if (r.ModelPath != null) { (modelObj, _) = await GfxModel.CreateObject(Source, r.ModelPath, true); GfxModel.PostObject(modelObj, obj.Position, obj.EulerAngles, obj.Scale, parent); }
        if (r.Record is ILigh ligh && GfxLight != null) {
            var s = GfxLight.CreateLight("Light", null, ligh.Radius, ligh.LightColor, cell.IsInterior);
            if (modelObj != null) GfxApi.Attach(GfxAttach.Find, s, modelObj, "AttachLight");
            else GfxModel.PostObject(s, obj.Position, obj.EulerAngles, obj.Scale, parent);
        }
    }

    /// <summary>
    /// Creates terrain representing a LAND record.
    /// </summary>
    async Task CreateLand(ILand land, Object parent) {
        const int LAND_SIDELENGTH_IN_SAMPLES = 65;
        const int VHGTIncrementToUnits = 8;
        const int LAND_TEXTUREINDICES = 256;
        const int VTEX_ROWS = 16;
        const int VTEX_COLUMNS = VTEX_ROWS;

        var heights = land.Heights;
        if (heights == null) return;

        // Read in the heights in Morrowind units.
        var newHeights = new float[LAND_SIDELENGTH_IN_SAMPLES, LAND_SIDELENGTH_IN_SAMPLES];
        var rowOffset = land.HeightOffset;
        for (var y = 0; y < LAND_SIDELENGTH_IN_SAMPLES; y++) {
            rowOffset += heights[y * LAND_SIDELENGTH_IN_SAMPLES];
            newHeights[y, 0] = rowOffset * VHGTIncrementToUnits;
            var colOffset = rowOffset;
            for (var x = 1; x < LAND_SIDELENGTH_IN_SAMPLES; x++) {
                colOffset += heights[(y * LAND_SIDELENGTH_IN_SAMPLES) + x];
                newHeights[y, x] = colOffset * VHGTIncrementToUnits;
            }
        }

        // Change the heights to percentages.
        newHeights.GetExtrema(out var minHeight, out var maxHeight);
        for (var y = 0; y < LAND_SIDELENGTH_IN_SAMPLES; y++)
            for (var x = 0; x < LAND_SIDELENGTH_IN_SAMPLES; x++)
                newHeights[y, x] = System.Polyfill.ChangeRange(newHeights[y, x], minHeight, maxHeight, 0f, 1f);

        // Texture the terrain.
        var textureManager = GfxModel.TextureManager;
        var indexs = land.Vtex ?? new uint[LAND_TEXTUREINDICES]; var layers = new List<GfxTerrainLayer<Texture>>(); var layerIndexs = new Dictionary<int, int>();
        for (var i = 0; i < indexs.Length; i++) {
            var index = (int)(indexs[i] - 1);
            if (layerIndexs.ContainsKey(index)) continue;
            // Load terrain texture.
            var path = index >= 0 ? Query.FindLtex(index).Path : DefaultLandTexturePath;
            var (tex, _) = await GfxModel.CreateTexture(Source, path);
            if (!TerrainLayers.TryGetValue(tex, out var layer)) {
                layer = new GfxTerrainLayer<Texture> {
                    Texture = tex,
                    Smoothness = .3f,
                    Metallic = .2f,
                    Specular = Color.Black,
                    TileSize = new Vector2(6, 6)
                };
                layer.MaskMapTexture = textureManager.CreateSolidTexture(1, 1, [layer.Metallic, 0f, 0f, layer.Smoothness]);
                layer.NormalMapTexture = textureManager.CreateNormalMapTexture(tex);
                TerrainLayers[tex] = layer;
            }
            var layerIndex = layers.Count; layers.Add(layer); layerIndexs.Add(index, layerIndex);
        }
        var newLayers = layers.ToArray();

        // Create the alpha map.
        var alphaMap = new float[VTEX_ROWS, VTEX_COLUMNS, newLayers.Length];
        for (var y = 0; y < VTEX_ROWS; y++) {
            int yMajor = y / 4, yMinor = y - (yMajor * 4);
            for (var x = 0; x < VTEX_COLUMNS; x++) {
                int xMajor = x / 4, xMinor = x - (xMajor * 4);
                var texIndex = (int)indexs[(yMajor * 64) + (xMajor * 16) + (yMinor * 4) + xMinor] - 1;
                alphaMap[y, x, texIndex >= 0 ? layerIndexs[texIndex] : 0] = 1;
            }
        }

        // Create the terrain.
        var heightRange = (maxHeight - minHeight) / MeterInUnits;
        var position = new Vector3(land.GridId.X * CellLengthInMeters, land.GridId.Y * CellLengthInMeters, minHeight / MeterInUnits);
        var sampleDistance = CellLengthInMeters / (LAND_SIDELENGTH_IN_SAMPLES - 1);
        var terrainData = GfxTerrain.CreateTerrainData(0, newHeights, heightRange, sampleDistance, newLayers, alphaMap);
        GfxTerrain.CreateTerrain("terrain", position, terrainData, parent);
    }

    void CreateReflectionProbe(ICell cell, Object parent) {
        if (cell.IsInterior) return;
        var gridId = cell.GridId;
        var position = new Vector3(gridId.X * CellLengthInMeters, 0f, gridId.Y * CellLengthInMeters);
        GfxLight.CreateReflectionProbe("probe", position, parent);
    }
}

#endregion

//List<string> GetLandTextures(ILand land) {
//    var vtex = land.Vtex;
//    if (land.Heights == null || vtex == null) return null;
//    var paths = new List<string>();
//    var indexs = vtex.Distinct().ToArray();
//    for (var i = 0; i < indexs.Length; i++) {
//        var index = (short)indexs[i] - 1;
//        if (index < 0) { paths.Add(DefaultLandTexturePath); continue; }
//        var ltex = Query.FindLtex(index);
//        paths.Add(ltex.Path);
//    }
//    return paths;
//}

//public override IEnumerator CellCoroutine(ICell cell, ILand land, object obj, object objectsObj) {
//    // Start pre-loading all required textures for the terrain.
//    //if (land != null) {
//    //    var landTextures = GetLandTextures(land);
//    //    //if (landTextures != null)
//    //    //    foreach (var landTexture in landTextures) GfxModel.PreloadTexture(landTexture);
//    //    yield return null;
//    //}

//    // Start pre-loading all required files for referenced objects. The NIF manager will load the textures as well.
//    //foreach (var s in cellRefs)
//    //    if (s.ModelPath != null) GfxModel.PreloadObject(s.ModelPath);
//    //yield return null;

//    // Extract information about referenced objects.
//    var cellRefs = GetCellRefs(cell);

//    // Instantiate terrain.
//    if (land != null) {
//        var task = LandCoroutine(land, (Object)obj);
//        while (task.MoveNext()) yield return null;
//        yield return null;
//    }

//    // Instantiate objects.
//    foreach (var s in cellRefs) CellObject(cell, (Object)objectsObj, s);
//}

File: rust/core/openstack/src/manager.rs
// PORT-SOURCE: Core/OpenStack/Manager.cs
// PORT-SHA: d562394db202ebac
// PORT-STATUS: done
//
// World streaming: `CellManager` keeps a window of cells loaded around the
// player, builds each through a `CellBuilder`, and cancels/destroys cells that
// fall out of range. `CellBuilder<Object, Material, Texture, Shader>` is the
// generic that closes over a backend's handle types — the same shape `gfx` uses,
// so this reuses `openstack_gfx::gfx::Backend`'s associated types.
//
// ===================== FOUR C#-SIDE BUGS ==================================
//
//   1. **`BeginCellByName` files every named cell under `Int3.Zero`.**
//
//          var cell = BuildCell(record); Cells[Int3.Zero] = cell;
//
//      So loading two cells by name silently evicts the first — the dictionary
//      key has nothing to do with the cell. Worse, `UpdateCells` then measures
//      Chebyshev distance from `(0,0,0)` for that entry, so a named interior
//      cell is destroyed or hidden based on a coordinate it never had. The port
//      keys by the record's own `grid_id`.
//
//   2. **`modelObj != null` is always true for a value-type backend.** `Object`
//      is an *unconstrained* type parameter, so `Object modelObj = default;`
//      followed by `if (modelObj != null)` compares a boxed value against null.
//      If a backend uses a struct handle — a Unity instance id, a `u32` GPU
//      index, anything not a class — the check passes even when the model
//      failed to load, and `GfxApi.Attach` runs against `default`. Rust's
//      `Option<B::Object>` makes the two states distinct.
//
//   3. **`TerrainLayers` is `static` on a generic class.** Same defect as
//      `gfx`'s `TextureManager`: one cache per closed generic type, shared by
//      every instance, never evicted. Two builders over different sources hand
//      each other's terrain layers back. It is an instance field here.
//
//   4. **`UpdateCells` scans the whole square once per ring.** The loop is
//      `for r in 0..=radius { for x in minX..=maxX { for y in minY..=maxY {
//      if (d == r) ... } } }`, so it makes `radius + 1` full passes over
//      `(2*radius+1)^2` positions to visit each once — 726 iterations for
//      radius 5 where 121 would do. Correct, just wasteful; the port walks each
//      ring directly and keeps the same near-to-far ordering.
//
// Also: `GfxApi`/`GfxModel`/`GfxLight`/`GfxTerrain` are initialised by unchecked
// casts on array indices (`(IOpenGfxApi<..>)gfx[GfX.XApi]`), so a short or
// mis-ordered `gfx` array throws `IndexOutOfRangeException` or
// `InvalidCastException` at construction. They are typed `Option` fields here.

use std::collections::HashMap;

use openstack_gfx::gfx::{Backend, GfxAttach};
use openstack_polyio::system_numerics::polyfill::Int3;
use openstack_polyio::system_numerics::vector3::Vec3;

/// C# `CellManager.ICellXref` — one placed object in a cell.
pub trait CellXref {
    fn name(&self) -> &str;
    fn scale(&self) -> Option<f32>;
    fn position(&self) -> Vec3;
    fn euler_angles(&self) -> Vec3;
}

/// C# `CellManager.ICellXrefModel`.
pub trait CellXrefModel {
    fn model_path(&self) -> &str;
}

/// C# `CellManager.ICell`.
pub trait Cell {
    fn id(&self) -> u32;
    fn is_interior(&self) -> bool;
    fn grid_id(&self) -> Int3;
    fn name(&self) -> &str;
    fn ambient_light(&self) -> Option<u32>;
    fn xrefs(&self) -> &[Box<dyn CellXref>];
}

/// C# `CellManager.ILand` — terrain for one exterior cell.
pub trait Land {
    fn grid_id(&self) -> Int3;
    fn vtex(&self) -> &[u32];
    fn height_offset(&self) -> f32;
    fn heights(&self) -> Option<&[i8]>;
}

/// C# `CellManager.ILtex`.
pub trait Ltex {
    fn intv(&self) -> i64;
    fn path(&self) -> &str;
}

/// C# `CellManager.ILigh`.
pub trait Ligh {
    fn radius(&self) -> f32;
    fn light_color(&self) -> u32;
}

/// C# `CellManager.IQuery` — the world database.
pub trait Query {
    fn meter_in_units(&self) -> f32;
    fn cell_length_in_meters(&self) -> f32;
    /// C# `Radius` — `[0]` is the load radius, `[1]` the visible radius.
    ///
    /// The C# indexes this array directly in the constructor
    /// (`query.Radius[0]`, `query.Radius[1]`), so a shorter array throws there.
    /// Split into two methods so the shape cannot be wrong.
    fn load_radius(&self) -> i32;
    fn visible_radius(&self) -> i32;
    fn world(&self) -> i32;
    fn set_world(&mut self, world: i32);
    fn get_cell_id(&self, point: Vec3) -> Int3;
    fn find_cell(&self, cell: Int3) -> Option<Box<dyn Cell>>;
    fn find_cell_by_name(&self, name: &str) -> Option<Box<dyn Cell>>;
    fn find_land(&self, cell: Int3) -> Option<Box<dyn Land>>;
    fn find_ltex(&self, index: i32) -> Option<Box<dyn Ltex>>;
}

/// C# `abstract class CellBuilder`.
///
/// The C# pairs an untyped abstract base (`(object, object) CreateContainers`)
/// with a generic subclass that casts back — an artefact of needing a
/// non-generic handle for `CellManager` to hold. Rust parameterises
/// `CellManager` by `B: Backend` instead, so one trait suffices and no cast
/// is needed anywhere.
pub trait CellBuilder<B: Backend> {
    /// C# `CreateContainers(string name)` -> `(cell object, objects object)`.
    fn create_containers(&mut self, name: &str) -> (B::Object, B::Object);

    /// C# `SetVisible(object src, bool visible)`.
    fn set_visible(&mut self, src: &B::Object, visible: bool);

    /// C# `Destroy(object src)`.
    fn destroy(&mut self, src: &B::Object);

    /// C# `Coroutine(...)` — one step of cell construction; `false` when done.
    ///
    /// The C# returns `IAsyncEnumerator<object>` and the queue awaits
    /// `MoveNextAsync`. Kept synchronous and steppable: the async-runtime
    /// decision is still open (see PORTING.md), and cell building is CPU work
    /// interleaved with the frame, not I/O waiting.
    fn step(&mut self, handle: &mut CellBuild<B>) -> bool;

    /// C# `GfxAttach` pass-through, used when attaching a light to a model.
    fn attach(&mut self, _mode: GfxAttach, _child: &B::Object, _parent: &B::Object) {}
}

/// In-progress construction state for one cell. Replaces the C#'s
/// `IAsyncEnumerator<object>` handle.
pub struct CellBuild<B: Backend> {
    pub obj: B::Object,
    pub objects_obj: B::Object,
    pub grid_id: Int3,
    pub is_interior: bool,
    /// How many `step` calls have run. Builders use this to resume.
    pub progress: usize,
    pub done: bool,
}

/// C# `CellManager.Cell`.
pub struct LoadedCell<B: Backend> {
    pub build: CellBuild<B>,
    pub record: Box<dyn Cell>,
}

/// C# `class CellManager(IQuery, AsyncCoroutineQueue, CellBuilder)`.
pub struct CellManager<B: Backend, Q: Query, CB: CellBuilder<B>> {
    query: Q,
    builder: CB,
    cells: HashMap<Int3, LoadedCell<B>>,
    radius: i32,
    radius2: i32,
}

impl<B: Backend, Q: Query, CB: CellBuilder<B>> CellManager<B, Q, CB> {
    pub fn new(query: Q, builder: CB) -> Self {
        let radius = query.load_radius();
        let radius2 = query.visible_radius();
        Self { query, builder, cells: HashMap::new(), radius, radius2 }
    }

    pub fn query(&self) -> &Q {
        &self.query
    }

    pub fn loaded_len(&self) -> usize {
        self.cells.len()
    }

    /// C# `BeginCell(Int3 point)`.
    pub fn begin_cell(&mut self, point: Int3) -> bool {
        let Some(record) = self.query.find_cell(point) else {
            return false;
        };
        let cell = self.build_cell(record);
        self.cells.insert(point, cell);
        true
    }

    /// C# `BeginCellByName(string name)`.
    ///
    /// Keys by the record's own `grid_id`, not `Int3::ZERO` — see bug 1.
    pub fn begin_cell_by_name(&mut self, name: &str) -> bool {
        let Some(record) = self.query.find_cell_by_name(name) else {
            return false;
        };
        let key = record.grid_id();
        let cell = self.build_cell(record);
        self.cells.insert(key, cell);
        true
    }

    /// C# `BuildCell(ICell cell)`.
    fn build_cell(&mut self, record: Box<dyn Cell>) -> LoadedCell<B> {
        // C# `Debug.Assert(cell != null)` — unrepresentable here.
        let name = if record.is_interior() {
            format!("cell {}", record.name())
        } else {
            format!("cell {}", record.grid_id())
        };
        let (obj, objects_obj) = self.builder.create_containers(&name);
        LoadedCell {
            build: CellBuild {
                obj,
                objects_obj,
                grid_id: record.grid_id(),
                is_interior: record.is_interior(),
                progress: 0,
                done: false,
            },
            record,
        }
    }

    /// C# `UpdateCells(Vector3 position, bool immediate, int radius)`.
    ///
    /// Walks rings outward from the player so nearby cells load first, visiting
    /// each position once — see bug 4.
    pub fn update_cells(&mut self, position: Vec3, immediate: bool, radius: Option<i32>) {
        let radius = radius.unwrap_or(self.radius);
        let point = self.query.get_cell_id(position);
        let world = self.query.world();

        // Evict anything outside the square, collecting first so the map is not
        // mutated mid-iteration (the C# does the same).
        let out_of_range: Vec<Int3> = self
            .cells
            .keys()
            .filter(|s| {
                (s.x - point.x).abs() > radius || (s.y - point.y).abs() > radius
            })
            .copied()
            .collect();
        for s in out_of_range {
            self.destroy_cell(s);
        }

        // Ring by ring, nearest first.
        for r in 0..=radius {
            for p in ring(point, r, world) {
                if self.cells.contains_key(&p) {
                    continue;
                }
                if self.begin_cell(p) && immediate {
                    self.finish_cell(p);
                }
            }
        }

        // Visibility by Chebyshev distance.
        let keys: Vec<Int3> = self.cells.keys().copied().collect();
        for p in keys {
            let d = (point.x - p.x).abs().max((point.y - p.y).abs());
            let visible = d <= self.radius2;
            if let Some(c) = self.cells.get(&p) {
                let obj = &c.build.objects_obj;
                // Clone the handle so the builder borrow does not alias the map.
                let obj = obj.clone();
                self.builder.set_visible(&obj, visible);
            }
        }
    }

    /// Advance every in-progress cell by one step. Replaces the C#'s
    /// `AsyncCoroutineQueue.Run` for this workload.
    ///
    /// Each cell is removed from the map, stepped, and reinserted. That is what
    /// keeps `&mut self.builder` from aliasing `&mut self.cells`, and it also
    /// means a builder is free to look at the manager's other cells if it ever
    /// needs to.
    pub fn step_cells(&mut self) -> usize {
        let keys: Vec<Int3> = self
            .cells
            .iter()
            .filter(|(_, c)| !c.build.done)
            .map(|(k, _)| *k)
            .collect();
        let mut stepped = 0;
        for k in keys {
            if let Some(mut c) = self.cells.remove(&k) {
                let more = self.builder.step(&mut c.build);
                c.build.done = !more;
                self.cells.insert(k, c);
                stepped += 1;
            }
        }
        stepped
    }

    /// C# `Queue.WaitFor(cell.Task)` — run one cell to completion.
    pub fn finish_cell(&mut self, point: Int3) {
        let Some(mut c) = self.cells.remove(&point) else { return };
        while !c.build.done {
            let more = self.builder.step(&mut c.build);
            c.build.done = !more;
        }
        self.cells.insert(point, c);
    }

    /// C# `DestroyCell(Int3 point)`.
    ///
    /// Returns whether a cell was actually there. The C# logs
    /// `Log.Error("Tried to destroy a cell that is not created.")` on a miss —
    /// through the `Log` whose sink defaults to null and throws, per the
    /// `polyfills` findings.
    pub fn destroy_cell(&mut self, point: Int3) -> bool {
        match self.cells.remove(&point) {
            Some(c) => {
                self.builder.destroy(&c.build.obj);
                true
            }
            None => false,
        }
    }

    /// C# `DestroyAllCells()`.
    pub fn destroy_all_cells(&mut self) {
        let cells = std::mem::take(&mut self.cells);
        for (_, c) in cells {
            self.builder.destroy(&c.build.obj);
        }
    }
}

/// Positions at exactly Chebyshev distance `r` from `center`, in the `world` plane.
fn ring(center: Int3, r: i32, world: i32) -> Vec<Int3> {
    if r == 0 {
        return vec![Int3::new(center.x, center.y, world)];
    }
    let mut out = Vec::with_capacity((8 * r) as usize);
    for x in (center.x - r)..=(center.x + r) {
        out.push(Int3::new(x, center.y - r, world));
        out.push(Int3::new(x, center.y + r, world));
    }
    for y in (center.y - r + 1)..(center.y + r) {
        out.push(Int3::new(center.x - r, y, world));
        out.push(Int3::new(center.x + r, y, world));
    }
    out
}

/// C# `interface IDatabase`.
///
/// Both members take and return `object`, so nothing about the interface is
/// checkable. It has no implementors in the solution; ported as a generic trait
/// so an implementor states its own types.
pub trait Database {
    type Src;
    type Converted;
    type Queried;
    fn convert(&self, src: Self::Src) -> Self::Converted;
    fn query(&self, src: Self::Src) -> Self::Queried;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ring_zero_is_the_center_only() {
        let c = Int3::new(3, 4, 0);
        assert_eq!(ring(c, 0, 7), vec![Int3::new(3, 4, 7)]);
    }

    #[test]
    fn ring_has_eight_r_positions_and_all_at_distance_r() {
        let c = Int3::new(0, 0, 0);
        for r in 1..=4 {
            let ps = ring(c, r, 0);
            assert_eq!(ps.len(), (8 * r) as usize, "ring {r} size");
            for p in &ps {
                let d = p.x.abs().max(p.y.abs());
                assert_eq!(d, r, "{p:?} not at distance {r}");
            }
        }
    }

    #[test]
    fn rings_cover_the_square_exactly_once() {
        // This is the property the C#'s radius+1 nested passes achieve by
        // brute force; verify the direct walk matches.
        let c = Int3::new(2, -3, 0);
        let radius = 4;
        let mut seen = std::collections::HashSet::new();
        for r in 0..=radius {
            for p in ring(c, r, 0) {
                assert!(seen.insert((p.x, p.y)), "{p:?} visited twice");
            }
        }
        let expected = ((2 * radius + 1) * (2 * radius + 1)) as usize;
        assert_eq!(seen.len(), expected, "must cover the full square");
    }

    #[test]
    fn rings_are_ordered_nearest_first() {
        let c = Int3::new(0, 0, 0);
        let mut last = -1;
        for r in 0..=3 {
            for p in ring(c, r, 0) {
                let d = p.x.abs().max(p.y.abs());
                assert!(d >= last, "ordering regressed");
                last = d;
            }
        }
    }

    #[test]
    fn ring_uses_the_world_plane_for_z() {
        for p in ring(Int3::new(0, 0, 0), 2, 9) {
            assert_eq!(p.z, 9);
        }
    }
}
